### ISSUE TITLE
models are the same after loading lora parameters using peft library

### ISSUE THREAD

**Issue by ra-MANUJ-an** [huggingface/peft#1389]
I created a lora and tried to merge it with base model but somehow the new model and the original model is giving the same logits.

base_model is as follows:
```
LlamaForCausalLM(
  (model): LlamaModel(
    (embed_tokens): Embedding(32000, 4096)
    (layers): ModuleList(
      (0-31): 32 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=4096, out_features=4096, bias=False)
          (k_proj): Linear(in_features=4096, out_features=4096, bias=False)
          (v_proj): Linear(in_features=4096, out_features=4096, bias=False)
          (o_proj): Linear(in_features=4096, out_features=4096, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=4096, out_features=11008, bias=False)
          (up_proj): Linear(in_features=4096, out_features=11008, bias=False)
          (down_proj): Linear(in_features=11008, out_features=4096, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm()
        (post_attention_layernorm): LlamaRMSNorm()
      )
    )
    (norm): LlamaRMSNorm()
  )
  (lm_head): Linear(in_features=4096, out_features=32000, bias=False)
)

```
and the lora_model is created by following code:

```python

expert_lora_path = '/lora-llm/llama-2-7b-expert-shakespeare'
expert_lora_config = LoraConfig.from_pretrained(expert_lora_path)
expert_peft_model = PeftModel.from_pretrained(base_model, expert_lora_path, device_map='cuda').to('cuda')

```

and is as follows:

```
PeftModelForCausalLM(
  (base_model): LoraModel(
    (model): LlamaForCausalLM(
      (model): LlamaModel(
        (embed_tokens): Embedding(32000, 4096)
        (layers): ModuleList(
          (0-31): 32 x LlamaDecoderLayer(
            (self_attn): LlamaSdpaAttention(
              (q_proj): lora.Linear(
                (base_layer): Linear(in_features=4096, out_features=4096, bias=False)
                (lora_dropout): ModuleDict(
                  (default): Dropout(p=0.1, inplace=False)
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=4096, out_features=32, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=32, out_features=4096, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
              (k_proj): Linear(in_features=4096, out_features=4096, bias=False)
              (v_proj): lora.Linear(
                (base_layer): Linear(in_features=4096, out_features=4096, bias=False)
                (lora_dropout): ModuleDict(
                  (default): Dropout(p=0.1, inplace=False)
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=4096, out_features=32, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=32, out_features=4096, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
              (o_proj): Linear(in_features=4096, out_features=4096, bias=False)
              (rotary_emb): LlamaRotaryEmbedding()
            )
            (mlp): LlamaMLP(
              (gate_proj): Linear(in_features=4096, out_features=11008, bias=False)
              (up_proj): Linear(in_features=4096, out_features=11008, bias=False)
              (down_proj): Linear(in_features=11008, out_features=4096, bias=False)
              (act_fn): SiLU()
            )
            (input_layernorm): LlamaRMSNorm()
            (post_attention_layernorm): LlamaRMSNorm()
          )
        )
        (norm): LlamaRMSNorm()
      )
      (lm_head): Linear(in_features=4096, out_features=32000, bias=False)
    )
  )
)


```

even though, i can clearly see lora modules being injected into the base_model, the logits still remain the same.

I double checked the above argument by comparing the parameters of two models, by the following code:

```python

flag = True
for p1, p2 in zip(base_model.parameters(), antiexpert_peft_model.parameters()):
    if p1.data.ne(p2.data).sum() > 0:
         flag = False
print (flag)

```
which gives me True as response. I'm confused as what's wrong in my implementation or was there any error while training.

**Comment by ra-MANUJ-an** [comment]
Hi,
@pacman100 
@BenjaminBossan 
Appreciate if you guys have suggestions?

**Comment by schwarzwalder93** [comment]
@ra-MANUJ-an I am also facing similar issue. Can you share how you resolved it ?

**Comment by ra-MANUJ-an** [comment]
Hi @schwarzwalder93 sure!

You can take a look in the following colab notebook:
https://colab.research.google.com/github/DanielWarfield1/MLWritingAndResearch/blob/main/LoRA.ipynb#scrollTo=vSZZjyp1mPoJ
You need to enable and disable adapter layers.

**Comment by simon-lund** [comment]
Yes, I don't know how it works for multiple adapters yet, but would be nice to set the one adapter as active adapter after loading it here:

https://github.com/huggingface/peft/blob/bfc102c0c095dc9094cdd3523b729583bfad4688/src/peft/peft_model.py#L354

According to the doc of `load_adapter` one has to use `set_adapter` to activate it:
https://github.com/huggingface/peft/blob/bfc102c0c095dc9094cdd3523b729583bfad4688/src/peft/peft_model.py#L659

**Comment by BenjaminBossan** [comment]
Thanks for reporting. I could not reproduce the issue of having to explicitly enable the adapter after loading it. Here is the code that I used:

```python
import torch
from peft import LoraConfig, get_peft_model, PeftModel
from transformers import AutoModelForCausalLM, AutoTokenizer

def generate(model, text):
    torch.manual_seed(0)
    inputs = tokenizer(text, return_tensors='pt')
    output = model.generate(
        input_ids=inputs['input_ids'],
        min_new_tokens=10,
        max_new_tokens=15,
    )
    print(tokenizer.decode(output[0]))

model_id = "bigscience/bloomz-560m"
torch.manual_seed(0)
model = AutoModelForCausalLM.from_pretrained(model_id)
tokenizer = AutoTokenizer.from_pretrained(model_id)
text = "Alice: I love cats. What do you love?\nBob:"

# generation from base model
generate(model, text)
# prints: Alice: I love cats. What do you love? Bob: I love cats. What do you love? Bob and Alice love cats.

# create a PEFT model
torch.manual_seed(0)
# by setting init_lora_weights=False, we ensure that the fresh LoRA is not a no-op
config = LoraConfig(task_type="CAUSAL_LM", r=16, init_lora_weights=False)
peft_model = get_peft_model(model, config)

# sanity check: The LoRA model generates a different output:
generate(peft_model, text)
# prints: Alice: I love cats. What do you love? Bob: I love cats. What do you love? Bob</s>

# save and load roundtrip
peft_model.save_pretrained("/tmp/issue-1389")
torch.manual_seed(0)
model = AutoModelForCausalLM.from_pretrained(model_id)
loaded = PeftModel.from_pretrained(model, "/tmp/issue-1389")

# generate with the loaded model:
generate(loaded, text)
# prints: Alice: I love cats. What do you love? Bob: I love cats. What do you love? Bob</s>
```

As you can see, the loaded model generates the same output as the LoRA model before saving and loading without the need to explicitly enable the adapter. We also have an extensive test suite that ensures this. Enabling the adapter should only be necessary if loading multiple adapters.

> I created a lora and tried to merge it with base model

In the code you show, there is no merging going on. What did you mean by that?

> I double checked the above argument by comparing the parameters of two models, by the following code:

This does not really work, because PEFT will mutate the base model in-place, so after you load the PEFT model, the base model is modified in the same fashion, so cannot detect any differences in parameters. You would have to keep a copy of the base model in memory to do this comparison.
